KAFKA-8584: Support of ByteBuffer for bytes field implemented.

nizhikov opened the pull request:
Support of the `java.nio.ByteBuffer` for bytes field implemented.
To generate code using 'ByteBuffer` one should set `useByteBuffer` flag to true.

`TestByteBufferDataTest` added to test the new feature.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

nizhikov commented:
@guozhangwang I implemented KAFKA-8584. Can you, please, do the review?

guozhangwang commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -316,7 +316,10 @@ private String fieldAbstractJavaType(FieldSpec field) {
         } else if (field.type().isString()) {
             return "String";
-        } else if (field.type().isBytes()) {
+        } else if (field.type().isBytes() && field.useByteBuffer()) {
```
nit: we could have the nested condition within `field.type().isBytes()`

nizhikov replied:
fixed.

guozhangwang commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -482,7 +486,9 @@ private String readFieldFromReadable(FieldType type) {
         } else if (type.isString()) {
             return "readable.readNullableString()";
-        } else if (type.isBytes()) {
+        } else if (type.isBytes() && useByteBuffer) {
```
Ditto here.

nizhikov replied:
Sorry, but I don't know what "Ditto" means :)

I assume, that you want `!useByteArray` be removed.
If yes, then fixed.

guozhangwang commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -612,7 +618,9 @@ private String readFieldFromStruct(FieldType type, String name) {
             return String.format("struct.getString(\"%s\")", name);
-        } else if (type.isBytes()) {
+        } else if (type.isBytes() && useByteArray) {
+            return String.format("struct.getBytes(\"%s\")", name);
```
Ditto here.

nizhikov replied:
Sorry, but I don't know what "Ditto" means :)

I assume, that you want `!useByteArray` be removed.
If yes, then fixed.

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/protocol/Writable.java:
```diff
@@ -38,6 +39,17 @@ default void writeNullableBytes(byte[] arr) {
+        if (buf == null) {
+            writeInt(-1);
+        } else {
+            writeBytes(buf.array());
```
Is it safe to always assume the byte buffer is backed by the array exactly? Theoretically the same array can be used for multiple byte buffer. Maybe we should just write from offset to limit?

nizhikov replied:
> Theoretically the same array can be used for multiple byte buffer.

We can have different implementation of `ByteBuffer`, also(`DirectByteBuffer`, for example).
I reworked this method and moved it to the `Writable` implementation.

guozhangwang reviewed (commented):
@nizhikov Thanks for the PR! Could you also add some unit tests?

Ping @abbccdda @cmccabe could you take a look at this PR?

nizhikov commented:
@guozhangwang 

> Could you also add some unit tests?

I've added `TestByteBufferDataTest.java`, already.
This test similar to `TestUUIDDataTest` that was used to test support of `UUID` type.

Do we need more test? If yes, can you clarify, what should I check additionally?

nizhikov commented:
Hello,

@abbccdda @cmccabe 

Tests are green. Can you please, join the review?

cmccabe commented:
Hi @nizhikov, this looks like a duplicate of https://github.com/apache/kafka/pull/7032 , right?

nizhikov commented:
> Hi @nizhikov, this looks like a duplicate of #7032 , right?

Looks like that.
But tests are green in my PR :)

I think it up to the community which PR should be reviewed and merged

I will write to the dev-list about that.

Thanks, for pointing this out.

cmccabe commented:
Hi @nizhikov, 

We need to redo the logic in ApiKeys that deals with when a Message is considered to retain a buffer reference.  Currently, it assumes that all messages that include `bytes` fields retain a reference.  However, in fact, only fields that are translated into `ByteBuffer` retain a reference.

There is a big difference between using an array and using a byte buffer here.  The BB keeps a reference to the underlying buffer.  So you have to, for example, not free the underlying buffer, or reuse it for something else.  This isn't really clear with the nomenclature used in the patch.  I'd suggest "zeroCopy" as a flag would be clearer than "useByteBuffer".

`writeNullableByteBuffer`: I actually would prefer not to integrate writing the length into this function, because how we write the length is going change as part of KIP-482.

`Bytes#EMPTY_BUF`: I actually think this belongs in `ByteUtils` rather than in `Bytes`.  The rest of the functions in `Bytes` concern arrays, not `ByteBuffers`.  `ByteUtils`, on the other hand, has always had utilities for dealing with `ByteBuffer` objects.

nizhikov commented:
@cmccabe 

Thanks, for the review.
I resolved your comments, please, review PR one more time.

> We need to redo the logic in ApiKeys that deals with when a Message is considered to retain a buffer reference.

1. `Type.BYTE_BUFFER` added.
2. `ApiKeys#retainsBufferReference` changed.
3. Test `TestByteBufferDataTest#shouldUseByteBufferFieldInSchema` added.

> I'd suggest "zeroCopy" as a flag would be clearer than "useByteBuffer".

Renamed.

> `Bytes#EMPTY_BUF`: I actually think this belongs in ByteUtils rather than in Bytes

Moved.

> `writeNullableByteBuffer`: I actually would prefer not to integrate writing the length into this function

I don't understand the proposed changes. Seems, other methods that deals with length use the same technique to write length. Please, clarify, what should be changed.

nizhikov commented:
Fails unrelated

nizhikov commented:
Failure unrelated.

Retest this, please.

nizhikov commented:
Failure unrelated.

Retest this, please.

nizhikov commented:
Hello, @abbccdda @cmccabe @guozhangwang 

Did you have a chance to review my changes?
Seems, I addressed all review comments.

nizhikov commented:
Hello, @cmccabe 

Do we still need this patch after 0de61a4683b92bdee803c51211c3277578ab3edf ?

nizhikov commented:
1. Fixed all merge conflicts
2. Test for nullable `zeroCopy` field added.

Please, review.

Seems, we can extend `zeroCopy` approach to all kind of byte fields: `BYTE`, `NULLABLE_BYTES`, `COMPACT_BYTES`, `COMPACT_NULLABLE_BYTES`.
Should I implement it in this PR?

nizhikov commented:
Failure unrelated.

Retest this, please.

cmccabe commented:
Hi @nizhikov, sorry for the slowness in reviews.  We have been busy.

Can you please rebase your changes on top of trunk, rather than merging trunk into your branch?  It's very difficult to follow which changes are yours and which come from the merge, when you do it the other way.  I think we will need to clean this up into a single branch without merges in order to make progress here.

In Type.java, I see that you introduce a BYTE_BUFFER type.  Let's not add this.  We already have BYTES and NULLABLE_BYTES.  We don't need another type since in the manually generated Struct.java code, we don't plan on using byte arrays anywhere.

Similarly, let's clean up `ApiKeys#retainsBufferReference` so that it is using a function automatically generated in `ApiMessageType` that tells us whether the message contains zero copy buffers or not.  We can't use the Type information for that since zero-copy and non-zero copy buffers should have the same type.

Rather than having a new `TestByteBuffer.json` file, let's just create another field in `SimpleExampleMessage.json` to test the byte buffers.  The more files we have, the longer tests take, and that's a drain on developer productivity.  We should test both the zero copy and non-zero copy versions of the bytes field.  Similarly, nullable byte buffers don't need their own test message file.

In `FieldSpec`, we should throw an exception if `zeroCopy` is set and the type of the field is not "bytes".

In `MessageDataGenerator`, `limit() - position()` can be better expressed as `remaining()`

nizhikov committed:
KAFKA-8584: reset branch to the trunk

cmccabe commented on generator/src/main/java/org/apache/kafka/message/SchemaGenerator.java:
```diff
@@ -270,22 +272,22 @@ private String fieldTypeToSchemaType(FieldType type,
                 return nullable ? "Type.COMPACT_NULLABLE_BYTES" : "Type.COMPACT_BYTES";
             } else {
-                return nullable ? "Type.NULLABLE_BYTES" : "Type.BYTES";
+                return zeroCopy ? "Type.BYTE_BUFFER" : (nullable ? "Type.NULLABLE_BYTES" : "Type.BYTES");
```
Like I commented earlier, whether zero copy is enabled does not affect the type.  Please get rid of Type.BYTE_BUFFER, since it is redundant.

nizhikov replied:
fixed.

cmccabe commented on generator/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -116,6 +119,8 @@ public FieldSpec(@JsonProperty("name") String name,
         this.tag = Optional.ofNullable(tag);
         checkTagInvariants();
+
+        this.zeroCopy = zeroCopy;
```
Please add a check that the type is "bytes" if zero copy is enabled.  Throw an exception if it is not

nizhikov replied:
fixed.

nizhikov committed:
KAFKA-8584: code review fixes.

nizhikov committed:
KAFKA-8584: nits.

nizhikov committed:
KAFKA-8584: test update.

nizhikov commented:
@cmccabe 

Thanks for the detailed review!
I resolved all your comments.

Please, take a look.

> Can you please rebase your changes on top of trunk

Done.

> In Type.java, I see that you introduce a BYTE_BUFFER type. Let's not add this.

Removed.

> let's just create another field in SimpleExampleMessage.json to test the byte buffers.

Done.

> we should throw an exception if zeroCopy is set and the type of the field is not "bytes".

Done.

> In MessageDataGenerator, limit() - position() can be better expressed as remaining()

Done.

nizhikov committed:
KAFKA-8584: checkstyle update.

cmccabe commented:
Thanks, @nizhikov .  LGTM.

nizhikov commented:
@cmccabe Thank you so much, for the helpful review and merge!
